armada/api/armada_controller.py

```diff
@@ -16,46 +16,61 @@
 from armada.conf import default
 
 import json
-from falcon import HTTP_200
+import falcon
+from falcon import HTTP_200, HTTP_500
 
 from oslo_config import cfg
 from oslo_log import log as logging
 
 # Required Oslo configuration setup
 default.register_opts()
 
-from armada.handlers.armada import Armada as Handler
+from armada import api
+from armada.handlers.armada import Armada
 
 LOG = logging.getLogger(__name__)
 CONF = cfg.CONF
 DOMAIN = "armada"
 
 logging.setup(CONF, DOMAIN)
 
-class Apply(object):
+class Apply(api.BaseResource):
     '''
     apply armada endpoint service
     '''
 
     def on_post(self, req, resp):
 
-        # Load data from request and get options
-        data = json.load(req.stream)
-        opts = data['options']
+        try:
+            policy_action = 'chart_installer:install'
+            ctx = req.contex
 
-        # Encode filename
-        data['file'] = data['file'].encode('utf-8')
+            if not self.check_policy(policy_action, ctx):
+                self.access_denied(req, resp, policy_action)
+                return
 
-        armada = Handler(open('../../' + data['file']),
-                         disable_update_pre=opts['disable_update_pre'],
-                         disable_update_post=opts['disable_update_post'],
-                         enable_chart_cleanup=opts['enable_chart_cleanup'],
-                         dry_run=opts['dry_run'],
-                         wait=opts['wait'],
-                         timeout=opts['timeout'])
+            data = self.req_json(req)
+            opts = data['options']
+            if data is not None:
+                armada = Armada(
+                    open(data),
+                    disable_update_pre=opts.get('disable_update_pre', False),
+                    disable_update_post=opts.get('disable_update_post', False),
+                    enable_chart_cleanup=opts.get('enable_chart_cleanup', False),
+                    dry_run=opts.get('dry_run', False),
+                    wait=opts.get('wait', False),
+                )
 
-        armada.sync()
+                armada.sync()
 
-        resp.data = json.dumps({'message': 'Success'})
-        resp.content_type = 'application/json'
-        resp.status = HTTP_200
+            resp.data = json.dumps({'success': True})
+            resp.content_type = 'application/json'
+            resp.status = HTTP_200
+
+        except Exception:
+            self.error(req.contex, "Failed to apply manifest")
+            self.return_error(
+                resp,
+                HTTP_500,
+                message="Failed to apply manifest"
```

Comment:
Starting with sending a 500 for unexpected exceptions is okay, but you should drive to handling all exception types explicitly and providing the client with a informative error about why their request failed. If you know `Armada.sync()` will raise particular exceptions in certain circumstances, then you should handle each of those cases here with an information error message.